File: Source/code/forms/formCopyMultiple.pas
// ===================================================================
//
// (c) Paul Alan Freshney 2012-2022
// www.freshney.org :: paul@freshney.org :: maximumoctopus.com
//
// https://github.com/MaximumOctopus/LEDMatrixStudio
//
// Please do not modifiy this comment section

//
// ===================================================================

unit formCopyMultiple;

interface

uses
  Winapi.Windows, Winapi.Messages, System.SysUtils, System.Variants, System.Classes, Vcl.Graphics,
  Vcl.Controls, Vcl.Forms, Vcl.Dialogs, Vcl.StdCtrls, Vcl.ExtCtrls, Vcl.Buttons,

  languagehandler;

type
  TframeCopyMultiple = class(TForm)
    bOK: TBitBtn;
    bCancel: TBitBtn;
    Bevel1: TBevel;
    Label4: TLabel;
    Image1: TImage;
    GroupBox1: TGroupBox;
    Label1: TLabel;
    Label2: TLabel;
    Label3: TLabel;
    Label5: TLabel;
    Label6: TLabel;
    Label7: TLabel;
    Label8: TLabel;
    lWarningMessage: TLabel;
    eStartFrame: TEdit;
    eEndFrame: TEdit;
    eCopyTo: TEdit;
    cbSourceLayer: TComboBox;
    cbDestinationLayer: TComboBox;
    cbAllLayers: TCheckBox;
    procedure eStartFrameChange(Sender: TObject);
    procedure cbAllLayersClick(Sender: TObject);
    procedure FormShow(Sender: TObject);
    procedure FormCreate(Sender: TObject);
  private
    FFrameCount : integer;

    procedure SetGUILanguageText;

    function ValidateInputs: boolean;
  public
    { Public declarations }
  end;

  TCopyMultipleObject = record
    Process : boolean;
    StartFrame, EndFrame, CopyTo : integer;
    Source, Destination : integer;
    AllLayers : boolean;
  end;


function DoCopyMultiple(aFrameCount : integer; var aLayers : TStringList): TCopyMultipleObject;


var
  frameCopyMultiple : TFrameCopyMultiple;


implementation


{$R *.dfm}


function DoCopyMultiple(aFrameCount : integer; var aLayers : TStringList): TCopyMultipleObject;
var
  lLayer : integer;

begin
  with TFrameCopyMultiple.Create(Application) do
    try
      FFrameCount := aFrameCount;

      Result.Process := False;

      for lLayer := 0 to aLayers.Count - 1 do begin
        cbSourceLayer.Items.Add(aLayers[lLayer]);
        cbDestinationLayer.Items.Add(aLayers[lLayer]);
      end;

      cbSourceLayer.ItemIndex      := 0;
      cbDestinationLayer.ItemIndex := 0;

      // =======================================================================

      ShowModal;

      // =======================================================================

      if ModalResult = mrOK then begin
        Result.Process     := True;
        Result.StartFrame  := StrToInt(eStartFrame.Text);
        Result.EndFrame    := StrToInt(eEndFrame.Text);
        Result.CopyTo      := StrToInt(eCopyTo.Text);

        Result.Source      := cbSourceLayer.ItemIndex;
        Result.Destination := cbDestinationLayer.ItemIndex;

        Result.AllLayers   := cbAllLayers.Checked;
      end;

    finally
      Free;
    end;
end;


procedure TframeCopyMultiple.FormCreate(Sender: TObject);
begin
  SetGUILanguageText;
end;


procedure TframeCopyMultiple.FormShow(Sender: TObject);
begin
  eStartFrameChange(Nil);
end;


procedure TframeCopyMultiple.SetGUILanguageText;
begin
  Caption := GLanguageHandler.Text[kCopyMultipleFrames];

  Label1.Caption := GLanguageHandler.Text[kStart];
  Label2.Caption := GLanguageHandler.Text[kEnd];
  Label3.Caption := GLanguageHandler.Text[kCopyTo];
  Label5.Caption := GLanguageHandler.Text[kDestinationFramesMustExist];
  lWarningMessage.Caption := GLanguageHandler.Text[kWarningYouWillLoseData];
  Label7.Caption := GLanguageHandler.Text[kLayers];
  Label6.Caption := GLanguageHandler.Text[kSource];
  Label8.Caption := GLanguageHandler.Text[kDestination];
  cbAllLayers.Caption := GLanguageHandler.Text[kAllLayers];

  Label4.Caption := GLanguageHandler.Text[kWarningThisActionCannotBeUndone];

  bOK.Caption := GLanguageHandler.Text[kOK];
  bCancel.Caption := GLanguageHandler.Text[kCancel];
end;


procedure TframeCopyMultiple.cbAllLayersClick(Sender: TObject);
begin
  cbDestinationLayer.Enabled := not(cbAllLayers.Checked);

  eStartFrameChange(Nil);
end;


procedure TframeCopyMultiple.eStartFrameChange(Sender: TObject);
begin
  bOk.Enabled := ValidateInputs;
end;


function TFrameCopyMultiple.ValidateInputs: boolean;
var
  lStartFrame, lEndFrame, lCopyTo, lMaxFrames : integer;
  lWarning : boolean;

begin
  lWarning := False;

  lStartFrame := StrToIntDef(eStartFrame.Text, -1);
  lEndFrame   := StrToIntDef(eEndFrame.Text, -1);
  lCopyTo     := StrToIntDef(eCopyTo.Text, -1);
  lMaxFrames  := lCopyTo + (lEndFrame - lStartFrame);               // max frame to write to

  if cbAllLayers.Checked then begin
    if (lCopyTo >= lStartFrame) and (lCopyTo <= lEndFrame) then
      lWarning := True;
  end
  else begin
    if (lCopyTo >= lStartFrame) and (lCopyTo <= lEndFrame) and (cbSourceLayer.ItemIndex = cbDestinationLayer.ItemIndex) then
      lWarning := True;
  end;

  lWarningMessage.Visible := lWarning;

  Result := (lStartFrame <> -1) and (lEndFrame <> -1) and (lCopyTo <> -1) and
            (lStartFrame <= lEndFrame) and (lMaxFrames <= FFrameCount) and (lMaxFrames >= 1) and
            not(lWarning);
end;


end.

File: Source/Forms/FormCopyMultiple.cpp
// ===================================================================
//
//   (c) Paul Alan Freshney 2012-2024
//   www.freshney.org :: paul@freshney.org :: maximumoctopus.com
//
//   https://github.com/MaximumOctopus/LEDMatrixStudio
//
//   https://maximumoctopus.hashnode.dev/
//
//   C++ Rewrite October 11th 2023
//
// ===================================================================

#include <vcl.h>
#pragma hdrstop

#include "FormCopyMultiple.h"
#include "LanguageConstants.h"
#include "LanguageHandler.h"

extern LanguageHandler *GLanguageHandler;

//---------------------------------------------------------------------------
#pragma package(smart_init)
#pragma resource "*.dfm"
TfrmCopyMultiple *frmCopyMultiple;
//---------------------------------------------------------------------------

CopyMultipleObject OpenCopyMultiple(int frame_count, std::vector<std::wstring> &Layers)
{
	TfrmCopyMultiple *frmCopyMultiple = new TfrmCopyMultiple(Application);

	frmCopyMultiple->FrameCount = frame_count;

	CopyMultipleObject cmo;

	for (int layer = 0; layer < Layers.size(); layer++)
	{
		frmCopyMultiple->cbSourceLayer->Items->Add(Layers[layer].c_str());
		frmCopyMultiple->cbDestinationLayer->Items->Add(Layers[layer].c_str());
	}

	frmCopyMultiple->cbSourceLayer->ItemIndex      = 0;
	frmCopyMultiple->cbDestinationLayer->ItemIndex = 0;

	if (frmCopyMultiple->ShowModal() == mrOk)
	{
		cmo.Process     = true;
		cmo.StartFrame  = frmCopyMultiple->eStartFrame->Text.ToInt();
		cmo.EndFrame    = frmCopyMultiple->eEndFrame->Text.ToInt();
		cmo.CopyTo      = frmCopyMultiple->eCopyTo->Text.ToInt();

		cmo.Source      = frmCopyMultiple->cbSourceLayer->ItemIndex;
		cmo.Destination = frmCopyMultiple->cbDestinationLayer->ItemIndex;

		cmo.AllLayers   = frmCopyMultiple->cbAllLayers->Checked;
	}

	delete frmCopyMultiple;

	return cmo;
}

//---------------------------------------------------------------------------

__fastcall TfrmCopyMultiple::TfrmCopyMultiple(TComponent* Owner)
	: TForm(Owner)
{
	SetGuiLanguageText();
}


void __fastcall TfrmCopyMultiple::FormShow(TObject *Sender)
{
	eStartFrameChange(nullptr);
}


void __fastcall TfrmCopyMultiple::eStartFrameChange(TObject *Sender)
{
	bOK->Enabled = ValidateInputs();
}


void __fastcall TfrmCopyMultiple::cbAllLayersClick(TObject *Sender)
{
	cbDestinationLayer->Enabled = !cbAllLayers->Checked;

	eStartFrameChange(nullptr);
}


bool TfrmCopyMultiple::ValidateInputs()
{
	bool warning = false;

	int start = eStartFrame->Text.ToIntDef(-1);
	int end   = eEndFrame->Text.ToIntDef(-1);
	int copyto    = eCopyTo->Text.ToIntDef(-1);
	int maxframes = copyto + (end - start);               // max frame to write to

	if (cbAllLayers->Checked)
	{
		if (copyto >= start && copyto <= end)
		{
			warning = true;
		}
	}
	else
	{
		if (copyto >= start && copyto <= end && cbSourceLayer->ItemIndex == cbDestinationLayer->ItemIndex)
		{
			warning = true;
		}
	}

	lWarningMessage->Visible = warning;

	return (start != -1 && end != -1 && copyto != -1 &&
			start <= end && maxframes <= FrameCount && maxframes >= 1 && !warning);
}


void TfrmCopyMultiple::SetGuiLanguageText()
{
	Caption = GLanguageHandler->Text[kCopyMultipleFrames].c_str();

	Label1->Caption = GLanguageHandler->Text[kStart].c_str();
	Label2->Caption = GLanguageHandler->Text[kEnd].c_str();
	Label3->Caption = GLanguageHandler->Text[kCopyTo].c_str();
	Label5->Caption = GLanguageHandler->Text[kDestinationFramesMustExist].c_str();
	lWarningMessage->Caption = GLanguageHandler->Text[kWarningYouWillLoseData].c_str();
	Label7->Caption = GLanguageHandler->Text[kLayers].c_str();
	Label6->Caption = GLanguageHandler->Text[kSource].c_str();
	Label8->Caption = GLanguageHandler->Text[kDestination].c_str();
	cbAllLayers->Caption = GLanguageHandler->Text[kAllLayers].c_str();

	Label4->Caption = GLanguageHandler->Text[kWarningThisActionCannotBeUndone].c_str();

	bOK->Caption = GLanguageHandler->Text[kOK].c_str();
	bCancel->Caption = GLanguageHandler->Text[kCancel].c_str();
}
